## Setting up the files and getting an overview of its information

### Setup

In [481]:
import pandas as pd

In [482]:
data = pd.read_csv('../Data_prep/Louisville_Animal.csv')

In [483]:
data.head()

kennel animalid jurisdiction intype insubtype               indate  \
0  FOLLOWUP  A713651        40202  STRAY       OTC  2022/04/02 00:00:00   
1       N46  A713652        40208  STRAY       OTC  2022/04/02 00:00:00   
2       N17  A733898        40208  STRAY       OTC  2023/03/17 00:00:00   
3       N30  A712650        40222  STRAY       OTC  2022/03/16 00:00:00   
4       409  A743389        40216  STRAY       OTC  2023/08/14 00:00:00   

  surreason   outtype  outsubtype              outdate animaltype sex bites  \
0     STRAY       RTO   IN KENNEL  2022/04/05 00:00:00        DOG   F     N   
1     STRAY  ADOPTION     WALK IN  2022/04/07 00:00:00        CAT   N     N   
2     STRAY    FOSTER    VACATION  2023/04/12 00:00:00        DOG   S     N   
3     STRAY  ADOPTION    INTERNET  2022/03/29 00:00:00        DOG   N     Y   
4     STRAY      EUTH  BEHAV HIST  2023/08/23 00:00:00        DOG   M     Y   

  petsize              color              breed sourcezipcode  ObjectId  
0     MED        TAN / BLACK  ENG BULLDOG / MIX         40208         1  
1   LARGE          ORG TABBY        DOMESTIC SH         40208         2  
2     MED  CHOCOLATE / WHITE           PIT BULL         40208         3  
3   PUPPY                TAN     PIT BULL / MIX         40208         4  
4     MED         BL BRINDLE           PIT BULL         40216         5

In [484]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52606 entries, 0 to 52605
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   kennel         52606 non-null  object
 1   animalid       52606 non-null  object
 2   jurisdiction   40356 non-null  object
 3   intype         52606 non-null  object
 4   insubtype      52549 non-null  object
 5   indate         52606 non-null  object
 6   surreason      40356 non-null  object
 7   outtype        41974 non-null  object
 8   outsubtype     32056 non-null  object
 9   outdate        42002 non-null  object
 10  animaltype     52606 non-null  object
 11  sex            51732 non-null  object
 12  bites          40362 non-null  object
 13  petsize        50333 non-null  object
 14  color          52604 non-null  object
 15  breed          52579 non-null  object
 16  sourcezipcode  46975 non-null  object
 17  ObjectId       52606 non-null  int64 
dtypes: int64(1), object(17)
me

In [485]:
#get the data range
data['indate'].min(), data['indate'].max()

('2019/01/01 00:00:00', '2023/08/26 00:00:00')

### Cleaning

#### Checking for duplicates

In [486]:
#drop duplicates
data.drop_duplicates(inplace=True)

No duplicates

In [487]:
data.nunique()

kennel             501
animalid         43736
jurisdiction        37
intype              10
insubtype           28
indate            1699
surreason           60
outtype             15
outsubtype          64
outdate           1724
animaltype           7
sex                  5
bites                2
petsize              7
color              468
breed             1231
sourcezipcode      528
ObjectId         52606
dtype: int64

Notes from nunique -- 501 kennels, 43,736 unique animals (indicating that some appear mutliple times), 37 jurisdictions, ten types of intypes, with 28 subtypes, only 15 types of our with 64 out subtypes (may choose to only classify to the 15 for first iteration.)
Currently are 1231 breeds. may be worth looking at grouping/simplifying those. Also may be worth it to clean the animal type to just a few. 
5 sex types?

Subnotes after review of codebook: start with in and out types, then go from there but surrtypes is very important to keep specific. 5 sexes are male, female, spayed, neutured, and unknown. Spayed = female, neutured = male

In [488]:
#check info about animals in multiple times
data[data['animalid'].duplicated(keep=False)].sort_values(by='animalid')

kennel animalid jurisdiction      intype   insubtype  \
46298     N04  A332960        40209       STRAY       FIELD   
46922     301  A332960        40213       STRAY       FIELD   
49245  FOSTER  A341716        40216  CONFISCATE  OWNER DIED   
25479      NP  A341716        40223      FOSTER      RETURN   
19209     238  A358112        40213      FOSTER      RETURN   
...       ...      ...          ...         ...         ...   
4152   FOSTER  A743783        40218       STRAY         OTC   
3830      NF1  A743784        40218      FOSTER      RETURN   
3826   FOSTER  A743784        40218       STRAY         OTC   
4426      NF1  A743785        40218      FOSTER      RETURN   
4751   FOSTER  A743785        40218       STRAY         OTC   

                    indate   surreason   outtype  outsubtype  \
46298  2019/09/21 00:00:00       STRAY       RTO   IN KENNEL   
46922  2020/02/07 00:00:00       STRAY      EUTH   REQUESTED   
49245  2020/08/25 00:00:00  OWNER DIED    FOSTER     TOO OLD   
25479  2020/09/04 00:00:00  FOSTER RET  ADOPTION      FOSTER   
19209  2019/08/07 00:00:00       STRAY  TRANSFER  RESCUE GRP   
...                    ...         ...       ...         ...   
4152   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   
3830   2023/08/24 00:00:00       STRAY       NaN         NaN   
3826   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   
4426   2023/08/24 00:00:00       STRAY  ADOPTION    INTERNET   
4751   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   

                   outdate animaltype sex bites petsize               color  \
46298  2019/10/25 00:00:00        DOG   N     N   LARGE       BROWN / WHITE   
46922  2020/02/11 00:00:00        DOG   N     N   LARGE       BROWN / WHITE   
49245  2020/08/31 00:00:00        DOG   S     N     TOY                 RED   
25479  2020/09/04 00:00:00        DOG   S     N     TOY                 RED   
19209  2019/08/07 00:00:00        DOG   S     N     MED       BLACK / BLACK   
...                    ...        ...  ..   ...     ...                 ...   
4152   2023/08/21 00:00:00        CAT   N     N   KITTN  WHITE / GRAY TABBY   
3830                   NaN        CAT   N     N   KITTN          GRAY TABBY   
3826   2023/08/21 00:00:00        CAT   N     N   KITTN          GRAY TABBY   
4426   2023/08/26 00:00:00        CAT   S     N   KITTN                GRAY   
4751   2023/08/21 00:00:00        CAT   S     N   KITTN                GRAY   

                    breed sourcezipcode  ObjectId  
46298  AMER BULLDOG / MIX         40218     46299  
46922  AMER BULLDOG / MIX         40215     46923  
49245          POMERANIAN         40223     49246  
25479          POMERANIAN         40223     25480  
19209       LABRADOR RETR         40218     19210  
...                   ...           ...       ...  
4152          DOMESTIC SH         40218      4153  
3830          DOMESTIC SH         40218      3831  
3826          DOMESTIC SH         40218      3827  
4426          DOMESTIC SH         40218      4427  
4751          DOMESTIC SH         40218      4752  

[15907 rows x 18 columns]

#### Handling nulls

##### Setup

In [489]:
#display count of nulls for each col
data.isnull().sum()

kennel               0
animalid             0
jurisdiction     12250
intype               0
insubtype           57
indate               0
surreason        12250
outtype          10632
outsubtype       20550
outdate          10604
animaltype           0
sex                874
bites            12244
petsize           2273
color                2
breed               27
sourcezipcode     5631
ObjectId             0
dtype: int64

In [490]:
data.drop(['jurisdiction','kennel','sourcezipcode','outsubtype','animalid','ObjectId'], axis=1, inplace=True)

In [491]:
#drop animals with no outtype
data.dropna(subset=['outtype'], inplace=True)

In [492]:
#create a frame that contains all the rows with nulls
null_data = data[data.isnull().any(axis=1)]
null_data.head()

intype insubtype               indate surreason    outtype  \
473   LOST       WEB  2019/06/13 00:00:00       NaN   LOST EXP   
475  FOUND       WEB  2019/06/16 00:00:00       NaN  FOUND EXP   
495   LOST       WEB  2019/06/21 00:00:00       NaN   LOST EXP   
501  FOUND       WEB  2019/06/21 00:00:00       NaN  FOUND EXP   
503   LOST       WEB  2019/06/22 00:00:00       NaN   LOST EXP   

                 outdate animaltype sex bites petsize            color  \
473  2019/07/14 00:00:00        CAT   F   NaN     MED          UNKNOWN   
475  2019/07/17 00:00:00        DOG   M   NaN   SMALL          UNKNOWN   
495  2019/07/22 00:00:00        CAT   M   NaN   LARGE          UNKNOWN   
501  2019/07/22 00:00:00        CAT   M   NaN     NaN  UNKNOWN / WHITE   
503  2019/07/23 00:00:00        DOG   M   NaN     MED          UNKNOWN   

            breed  
473   DOMESTIC MH  
475  CHIHUAHUA LH  
495    MAINE COON  
501   DOMESTIC SH  
503        BEAGLE

In [493]:
data.isnull().sum()

intype           0
insubtype       56
indate           0
surreason     1797
outtype          0
outdate          9
animaltype       0
sex            145
bites         1797
petsize        402
color            2
breed            0
dtype: int64

##### Null Insubtype

In [494]:
#create a frame that contains all the insubtype nulls
null_insubtype = null_data[null_data['insubtype'].isnull()]
null_insubtype['intype'].value_counts()

STRAY         32
CONFISCATE    10
RETURN         9
FOSTER         4
OWNER SUR      1
Name: intype, dtype: int64

In [495]:
#fill nulls with 'Unknown'
data['insubtype'].fillna('Unknown', inplace=True)

##### Null surreason

In [496]:
null_surreason = null_data[null_data['surreason'].isnull()]
null_surreason['intype'].value_counts()

LOST     1259
FOUND     538
Name: intype, dtype: int64

In [497]:
data['surreason'].value_counts()

STRAY         31030
FOSTER RET     3253
ABANDON         801
DOA             569
RESPONSIBL      383
UNKNOWN         354
AGG PEOPLE      300
OTHER PET       278
INJURED         227
AGG ANIMAL      220
BITES           214
TOO MANY        201
OWNER MED       177
OWNER DIED      173
PETMEDICAL      167
EUTH MED        141
MOVE            123
VIOLATION       114
LANDLORD        113
CHILD PROB      111
HYPER           108
ILL             108
ALLERGIC        101
NO TIME          87
CRUELTY          87
EUTH BEHAV       74
TRANSFER         72
NO HOME          66
EUTH OLD         50
WILDLIFE         48
DESTRUC IN       47
AFRAID           42
COST             33
AGG FOOD         31
AGG FEAR         30
HOUSE SOIL       29
TOO YOUNG        28
ESCAPES          27
CHASES ANI       26
SHEDS            23
TOO BIG          20
EUTH OTHER       18
NOFRIENDLY       12
DISOBIDIEN       11
FOUND ANIM        7
JUMPS UP          6
NEW BABY          6
ATTENTION         6
VOCAL             4
NO PROTECT        4


In [498]:
#fill null surreason with "N/A"
data['surreason'].fillna('LOST/FOUND', inplace=True)

##### Null sex

In [499]:
data['sex'].value_counts()


N    17590
S    14924
M     4371
F     3717
U     1227
Name: sex, dtype: int64

In [500]:
data['sex'].fillna('U', inplace=True)

##### Null bites

In [501]:
data['bites'].fillna('U', inplace=True)

##### Null petsize

In [502]:
null_size = null_data[null_data['petsize'].isnull()]
null_size['animaltype'].value_counts()

DOG     276
CAT     125
BIRD      1
Name: animaltype, dtype: int64

In [503]:
null_size['breed'].value_counts()

DOMESTIC SH        60
PIT BULL           42
AMER SH            21
GERM SHEPHERD      18
DOMESTIC MH        18
                   ..
SCOT TERRIER        1
CHINESE SHARPEI     1
COLLIE SMOOTH       1
POODLE STND         1
ST BERNARD SMTH     1
Name: breed, Length: 78, dtype: int64

In [504]:
#check if any of the null size animals have / or mix in their breed
null_size[null_size['breed'].str.contains('/')]['breed'].value_counts()


BLACK/TAN HOUND    1
Name: breed, dtype: int64

In [505]:
null_size[null_size['breed'].str.contains('mix')]['breed'].value_counts()


Series([], Name: breed, dtype: int64)

In [506]:
#create a table of the mode petsize for each breed
mode_size = data.groupby('breed')['petsize'].agg(pd.Series.mode).to_frame()

In [507]:
#using the mode table, fill the nulls with the mode for each breed
data['petsize'].fillna(data['breed'].map(mode_size['petsize']), inplace=True)

In [508]:
#check for nulls in data
data.isnull().sum()

intype        0
insubtype     0
indate        0
surreason     0
outtype       0
outdate       9
animaltype    0
sex           0
bites         0
petsize       0
color         2
breed         0
dtype: int64

In [509]:
#drop the remaining nulls
data.dropna(inplace=True)

#### Creating new relevant columns

In [510]:
data

intype insubtype               indate  surreason   outtype  \
0           STRAY       OTC  2022/04/02 00:00:00      STRAY       RTO   
1           STRAY       OTC  2022/04/02 00:00:00      STRAY  ADOPTION   
2           STRAY       OTC  2023/03/17 00:00:00      STRAY    FOSTER   
3           STRAY       OTC  2022/03/16 00:00:00      STRAY  ADOPTION   
4           STRAY       OTC  2023/08/14 00:00:00      STRAY      EUTH   
...           ...       ...                  ...        ...       ...   
52601       STRAY       OTC  2019/09/21 00:00:00      STRAY  ADOPTION   
52602       STRAY     FIELD  2020/01/15 00:00:00      STRAY  TRANSFER   
52603  CONFISCATE   NEGLECT  2020/01/22 00:00:00  VIOLATION  ADOPTION   
52604  CONFISCATE   NEGLECT  2019/06/23 00:00:00    CRUELTY       RTO   
52605       STRAY       OTC  2019/09/20 00:00:00      STRAY       RTO   

                   outdate animaltype sex bites petsize              color  \
0      2022/04/05 00:00:00        DOG   F     N     MED        TAN / BLACK   
1      2022/04/07 00:00:00        CAT   N     N   LARGE          ORG TABBY   
2      2023/04/12 00:00:00        DOG   S     N     MED  CHOCOLATE / WHITE   
3      2022/03/29 00:00:00        DOG   N     Y   PUPPY                TAN   
4      2023/08/23 00:00:00        DOG   M     Y     MED         BL BRINDLE   
...                    ...        ...  ..   ...     ...                ...   
52601  2019/09/28 00:00:00        DOG   N     N   LARGE      BLACK / BROWN   
52602  2020/01/24 00:00:00        DOG   M     N     MED              BROWN   
52603  2020/02/15 00:00:00        DOG   S     N     MED      BLACK / WHITE   
52604  2019/06/27 00:00:00        DOG   M     N     MED               BLUE   
52605  2019/09/24 00:00:00        DOG   N     N     MED         BR BRINDLE   

                     breed  
0        ENG BULLDOG / MIX  
1              DOMESTIC SH  
2                 PIT BULL  
3           PIT BULL / MIX  
4                 PIT BULL  
...                    ...  
52601  LABRADOR RETR / MIX  
52602             PIT BULL  
52603  BORDER COLLIE / MIX  
52604             PIT BULL  
52605             PIT BULL  

[41963 rows x 12 columns]

##### Handling mixed breeds

In [511]:
#if the breed ends with " / MIX", contains ' / ', or contains 'MIX', then it is mixed
data['mixed'] = data['breed'].apply(lambda x: True if x.endswith(' / MIX') or ' / ' in x or 'MIX' in x else False)

In [512]:
data[data['mixed']]

intype insubtype               indate  surreason   outtype  \
0           STRAY       OTC  2022/04/02 00:00:00      STRAY       RTO   
3           STRAY       OTC  2022/03/16 00:00:00      STRAY  ADOPTION   
12          STRAY       OTC  2023/02/08 00:00:00      STRAY    FOSTER   
13          STRAY       OTC  2023/02/08 00:00:00      STRAY  TRANSFER   
15          STRAY       OTC  2023/05/25 00:00:00      STRAY  TRANSFER   
...           ...       ...                  ...        ...       ...   
52594       STRAY     FIELD  2020/01/15 00:00:00      STRAY      EUTH   
52595       STRAY     FIELD  2020/01/21 00:00:00      STRAY  ADOPTION   
52599  CONFISCATE    POLICE  2019/10/01 00:00:00      BITES      EUTH   
52601       STRAY       OTC  2019/09/21 00:00:00      STRAY  ADOPTION   
52603  CONFISCATE   NEGLECT  2020/01/22 00:00:00  VIOLATION  ADOPTION   

                   outdate animaltype sex bites petsize          color  \
0      2022/04/05 00:00:00        DOG   F     N     MED    TAN / BLACK   
3      2022/03/29 00:00:00        DOG   N     Y   PUPPY            TAN   
12     2023/02/16 00:00:00        DOG   N     N   SMALL    TAN / BLACK   
13     2023/02/25 00:00:00        DOG   N     N   SMALL    TAN / WHITE   
15     2023/05/31 00:00:00        DOG   M     N   PUPPY   GRAY / WHITE   
...                    ...        ...  ..   ...     ...            ...   
52594  2020/01/31 00:00:00        DOG   S     N     MED  WHITE / BLACK   
52595  2020/02/06 00:00:00        DOG   N     N     MED          BROWN   
52599  2020/04/17 00:00:00        DOG   M     N   LARGE  BLACK / BROWN   
52601  2019/09/28 00:00:00        DOG   N     N   LARGE  BLACK / BROWN   
52603  2020/02/15 00:00:00        DOG   S     N     MED  BLACK / WHITE   

                         breed  mixed  
0            ENG BULLDOG / MIX   True  
3               PIT BULL / MIX   True  
12         JACK RUSS TER / PUG   True  
13               BASENJI / MIX   True  
15     HAVANESE / TIBETAN TERR   True  
...                        ...    ...  
52594           PIT BULL / MIX   True  
52595           PIT BULL / MIX   True  
52599         ROTTWEILER / MIX   True  
52601      LABRADOR RETR / MIX   True  
52603      BORDER COLLIE / MIX   True  

[9311 rows x 13 columns]

In [513]:
#drop ' / MIX' from the end of the breed and replace with ''
data['breed'] = data['breed'].apply(lambda x: x.replace(' / MIX', ''))

In [514]:
#for breeds with multiple breeds, split the breeds into two columns
data[['breed','breed2']] = data['breed'].str.split(' / ', expand=True)
data[data['mixed']]

intype insubtype               indate  surreason   outtype  \
0           STRAY       OTC  2022/04/02 00:00:00      STRAY       RTO   
3           STRAY       OTC  2022/03/16 00:00:00      STRAY  ADOPTION   
12          STRAY       OTC  2023/02/08 00:00:00      STRAY    FOSTER   
13          STRAY       OTC  2023/02/08 00:00:00      STRAY  TRANSFER   
15          STRAY       OTC  2023/05/25 00:00:00      STRAY  TRANSFER   
...           ...       ...                  ...        ...       ...   
52594       STRAY     FIELD  2020/01/15 00:00:00      STRAY      EUTH   
52595       STRAY     FIELD  2020/01/21 00:00:00      STRAY  ADOPTION   
52599  CONFISCATE    POLICE  2019/10/01 00:00:00      BITES      EUTH   
52601       STRAY       OTC  2019/09/21 00:00:00      STRAY  ADOPTION   
52603  CONFISCATE   NEGLECT  2020/01/22 00:00:00  VIOLATION  ADOPTION   

                   outdate animaltype sex bites petsize          color  \
0      2022/04/05 00:00:00        DOG   F     N     MED    TAN / BLACK   
3      2022/03/29 00:00:00        DOG   N     Y   PUPPY            TAN   
12     2023/02/16 00:00:00        DOG   N     N   SMALL    TAN / BLACK   
13     2023/02/25 00:00:00        DOG   N     N   SMALL    TAN / WHITE   
15     2023/05/31 00:00:00        DOG   M     N   PUPPY   GRAY / WHITE   
...                    ...        ...  ..   ...     ...            ...   
52594  2020/01/31 00:00:00        DOG   S     N     MED  WHITE / BLACK   
52595  2020/02/06 00:00:00        DOG   N     N     MED          BROWN   
52599  2020/04/17 00:00:00        DOG   M     N   LARGE  BLACK / BROWN   
52601  2019/09/28 00:00:00        DOG   N     N   LARGE  BLACK / BROWN   
52603  2020/02/15 00:00:00        DOG   S     N     MED  BLACK / WHITE   

               breed  mixed        breed2  
0        ENG BULLDOG   True          None  
3           PIT BULL   True          None  
12     JACK RUSS TER   True           PUG  
13           BASENJI   True          None  
15          HAVANESE   True  TIBETAN TERR  
...              ...    ...           ...  
52594       PIT BULL   True          None  
52595       PIT BULL   True          None  
52599     ROTTWEILER   True          None  
52601  LABRADOR RETR   True          None  
52603  BORDER COLLIE   True          None  

[9311 rows x 14 columns]

In [515]:
#if mixed is true and breed2 is 'None' replace with 'Unknown'
data.loc[(data['mixed']) & (data['breed2'].isnull()), 'breed2'] = 'Unknown'

In [516]:
#fill nan in breed2 with 'N/A'
data['breed2'].fillna('N/A', inplace=True)

#### Creating date-connected columns

In [517]:
#create a new column called time_in that is indate - outdate
data['days_in'] = pd.to_datetime(data['outdate']) - pd.to_datetime(data['indate'])
data['days_in'] = data['days_in'].dt.days
data

intype insubtype               indate  surreason   outtype  \
0           STRAY       OTC  2022/04/02 00:00:00      STRAY       RTO   
1           STRAY       OTC  2022/04/02 00:00:00      STRAY  ADOPTION   
2           STRAY       OTC  2023/03/17 00:00:00      STRAY    FOSTER   
3           STRAY       OTC  2022/03/16 00:00:00      STRAY  ADOPTION   
4           STRAY       OTC  2023/08/14 00:00:00      STRAY      EUTH   
...           ...       ...                  ...        ...       ...   
52601       STRAY       OTC  2019/09/21 00:00:00      STRAY  ADOPTION   
52602       STRAY     FIELD  2020/01/15 00:00:00      STRAY  TRANSFER   
52603  CONFISCATE   NEGLECT  2020/01/22 00:00:00  VIOLATION  ADOPTION   
52604  CONFISCATE   NEGLECT  2019/06/23 00:00:00    CRUELTY       RTO   
52605       STRAY       OTC  2019/09/20 00:00:00      STRAY       RTO   

                   outdate animaltype sex bites petsize              color  \
0      2022/04/05 00:00:00        DOG   F     N     MED        TAN / BLACK   
1      2022/04/07 00:00:00        CAT   N     N   LARGE          ORG TABBY   
2      2023/04/12 00:00:00        DOG   S     N     MED  CHOCOLATE / WHITE   
3      2022/03/29 00:00:00        DOG   N     Y   PUPPY                TAN   
4      2023/08/23 00:00:00        DOG   M     Y     MED         BL BRINDLE   
...                    ...        ...  ..   ...     ...                ...   
52601  2019/09/28 00:00:00        DOG   N     N   LARGE      BLACK / BROWN   
52602  2020/01/24 00:00:00        DOG   M     N     MED              BROWN   
52603  2020/02/15 00:00:00        DOG   S     N     MED      BLACK / WHITE   
52604  2019/06/27 00:00:00        DOG   M     N     MED               BLUE   
52605  2019/09/24 00:00:00        DOG   N     N     MED         BR BRINDLE   

               breed  mixed   breed2  days_in  
0        ENG BULLDOG   True  Unknown        3  
1        DOMESTIC SH  False      N/A        5  
2           PIT BULL  False      N/A       26  
3           PIT BULL   True  Unknown       13  
4           PIT BULL  False      N/A        9  
...              ...    ...      ...      ...  
52601  LABRADOR RETR   True  Unknown        7  
52602       PIT BULL  False      N/A        9  
52603  BORDER COLLIE   True  Unknown       24  
52604       PIT BULL  False      N/A        4  
52605       PIT BULL  False      N/A        4  

[41963 rows x 15 columns]

In [518]:
#create two new columns in_season and out_season that are fall, winter, spring, summer based on the month of indate and outdate
data['in_season'] = pd.to_datetime(data['indate']).dt.month.map({1: 'WINTER', 2: 'WINTER', 3: 'SPRING', 4: 'SPRING', 5: 'SPRING', 6: 'SUMMER', 7: 'SUMMER', 8: 'SUMMER', 9: 'FALL', 10: 'FALL', 11: 'FALL', 12: 'WINTER'})
data['out_season'] = pd.to_datetime(data['outdate']).dt.month.map({1: 'WINTER', 2: 'WINTER', 3: 'SPRING', 4: 'SPRING', 5: 'SPRING', 6: 'SUMMER', 7: 'SUMMER', 8: 'SUMMER', 9: 'FALL', 10: 'FALL', 11: 'FALL', 12: 'WINTER'})

In [519]:
#drop indate and outdate columns
data.drop(columns=['indate', 'outdate'], inplace=True)

### Final review

In [520]:
data.head()

intype insubtype surreason   outtype animaltype sex bites petsize  \
0  STRAY       OTC     STRAY       RTO        DOG   F     N     MED   
1  STRAY       OTC     STRAY  ADOPTION        CAT   N     N   LARGE   
2  STRAY       OTC     STRAY    FOSTER        DOG   S     N     MED   
3  STRAY       OTC     STRAY  ADOPTION        DOG   N     Y   PUPPY   
4  STRAY       OTC     STRAY      EUTH        DOG   M     Y     MED   

               color        breed  mixed   breed2  days_in in_season  \
0        TAN / BLACK  ENG BULLDOG   True  Unknown        3    SPRING   
1          ORG TABBY  DOMESTIC SH  False      N/A        5    SPRING   
2  CHOCOLATE / WHITE     PIT BULL  False      N/A       26    SPRING   
3                TAN     PIT BULL   True  Unknown       13    SPRING   
4         BL BRINDLE     PIT BULL  False      N/A        9    SUMMER   

  out_season  
0     SPRING  
1     SPRING  
2     SPRING  
3     SPRING  
4     SUMMER

In [521]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41963 entries, 0 to 52605
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   intype      41963 non-null  object
 1   insubtype   41963 non-null  object
 2   surreason   41963 non-null  object
 3   outtype     41963 non-null  object
 4   animaltype  41963 non-null  object
 5   sex         41963 non-null  object
 6   bites       41963 non-null  object
 7   petsize     41963 non-null  object
 8   color       41963 non-null  object
 9   breed       41963 non-null  object
 10  mixed       41963 non-null  bool  
 11  breed2      41963 non-null  object
 12  days_in     41963 non-null  int64 
 13  in_season   41963 non-null  object
 14  out_season  41963 non-null  object
dtypes: bool(1), int64(1), object(13)
memory usage: 4.8+ MB


In [522]:
data.isna().sum()

intype        0
insubtype     0
surreason     0
outtype       0
animaltype    0
sex           0
bites         0
petsize       0
color         0
breed         0
mixed         0
breed2        0
days_in       0
in_season     0
out_season    0
dtype: int64

In [523]:
data.describe()

days_in
count  41963.000000
mean       9.528442
std       16.819386
min        0.000000
25%        1.000000
50%        4.000000
75%       11.000000
max      365.000000

In [524]:
#export to csv
data.to_csv('../Data_prep/Louisville_Animal_Cleaned.csv', index=False)